[concurrency] Change #isolated to mask out the TBI bits of the witness pointer of the implicit isolated any Actor pointer so we can do optimizations on TBI supporting platforms in the future.

gottesmm opened the pull request:
Specifically, this will let us perform an optimization in the future where a caller can set bits in the TBI byte to signal information dynamically to a nonisolated(nonsending) callee. The intent is that this will let us signal to a nonisolated(nonsending) callee that the caller isn't performing any isolated work before it hops (or potentially a caller of the caller) hops meaning that the nonisolated(nonsending) function can avoid a tail hop that restores isolation after calling some sort of actor isolated call.

E.x.:

```swift
nonisolated(nonsending) func callsActor(_ a: MyActor) async {
  await a.doSomething()
}

@concurrent func runOnBackground() async {}

actor A {
  func doSomething() async { ... }

  func caller() async {
    await callsActor(self)
    await runOnBackgroud()
  }
}
```

In the above case, the hop back in callsActor is not needed dynamically in caller since we are going to immediately await ontot he background. But we do not know that. An optimization can be written that can recognize this case and dynamically tell via the tbi bits of the protocol witness table of the any Actor call that the hop to executor call can avoid the hop.

Since this only applies to the implicit isolated any Actor parameter which is only used for the purpose of hopping, it is safe to do this since when we hop we just call the Actor unowned executor dispatch thunk which directly accesses the witness table by loading from that pointer... meaning that anything set in the TBI bit is irrelevent.

The only other way to grab this implicit bit is to use `#isolation` and that is where this commit comes into play. Since we can access the actor via `#isolation`, we need to ensure that we mask out the TBI bits so that the rest of the runtime/compiled code never see the TBI bits. This is because we do not want the fact that we are using these TBI bits to ever escape into other code (e.x.: consider things that may cast).

So we have basically created a hermetically sealed world where we can pass information from callee to caller using the protocol witness TBI bits.

rdar://156525771

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

gottesmm commented:
@swift-ci smoke test

gottesmm commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -7379,13 +7379,82 @@ RValue RValueEmitter::visitCurrentContextIsolationExpr(
-        isolatedMV = ManagedValue::forBorrowedRValue(isolatedArg);
-      } else {
-        isolatedMV = ManagedValue::forUnmanagedOwnedValue(isolatedArg);
+      auto isolatedMV = ManagedValue::forBorrowedRValue(isolatedArg);
```
It will never be unowned. We always know that it will be @guaranteed. I checked.

gottesmm commented on include/swift/Basic/LangOptions.h:
```diff
@@ -670,6 +670,9 @@ namespace swift {
 #endif
 
+    /// Set to true if we support AArch64TBI.
+    bool HasAArch64TBI = false;
```
The reason I didn't put it into IRGen options is that I am using it in SILGen.

gottesmm replied:
@DougGregor I was thinking about this. Really the whole feature should be in IRGen and just be a flag on an apply. That being said, I want to do that later on main.

gottesmm commented:
@swift-ci smoke test

ktoso reviewed (approved):
AFAICS looking good 👍

DougGregor commented on lib/SILGen/ConcurrencyUtils.cpp:
```diff
@@ -0,0 +1,163 @@
+ManagedValue swift::Lowering::clearImplicitIsolatedActorBits(
+    SILGenFunction &SGF, SILLocation loc, ManagedValue isolatedMV) {
+  auto &ctx = SGF.getASTContext();
+  if (!ctx.LangOpts.hasFeature(Feature::NonisolatedNonsendingDynamicHopElim) &&
```
This part shouldn't be behind the feature flag, right? It's the bit stripping we will always need to do.

DougGregor commented on lib/SILGen/ConcurrencyUtils.cpp:
```diff
@@ -0,0 +1,163 @@
+    SILGenFunction &SGF, SILLocation loc, ManagedValue isolatedMV) {
+  auto &ctx = SGF.getASTContext();
+  if (!ctx.LangOpts.hasFeature(Feature::NonisolatedNonsendingDynamicHopElim) &&
+      !ctx.LangOpts.HasAArch64TBI) {
```
Why are we checking for TBI here? It seems like we effectively need to distinguish between places where TBI will strip the bits for us (e.g., we're dereferencing the pointer) vs those where we treat the pointer as a value and we need to strip the bits ourselves (e.g., when building a runtime data structure involving this witness table).

jamieQ commented on test/Concurrency/attr_execution/conversions_silgen.swift:
```diff
@@ -152,14 +152,15 @@ public func testCallerToCaller(_ x: nonisolated(nonsending) @escaping () async -
 
```
@gottesmm  just out of curiosity, how did you perform all these changes to update the test expectations? were they automated in some manner?

gottesmm committed:
[concurrency] Add a new type Builtin.ImplicitActor.

This is currently not wired up to anything. I am going to wire it up in
subsequent commits.

The reason why we are introducing this new Builtin type is to represent that we
are going to start stealing bits from the protocol witness table pointer of the
Optional<any Actor> that this type is bitwise compatible with. The type will
ensure that this value is only used in places where we know that it will be
properly masked out giving us certainty that this value will not be used in any
manner without it first being bit cleared and transformed back to Optional<any
Actor>.

gottesmm committed:
[concurrency] Implement bit masking for TBI when available or in tagged pointer bits otherwise.

Specifically, when TBI is available we use the bottom two bits of the top nibble
(bits 60,61). On platforms without TBI, we use the bottom two tagged pointer
bits (bits 0, 1).

rdar://156525771

gottesmm committed:
[silgen] Use Builtin.ImplicitActor instead of Optional<any Actor> to represent the implicit isolated parameter.

NOTE: We are not performing any bitmasking at all now. This is so that we can
transition the code base/tests to expect Builtin.ImplicitActor instead
of Optional<any Actor>.

NOTE: The actual test changes are in the next commit. I did this to make it
easier to review the changes.

This should not have any user visible changes.

gottesmm committed:
[silgen] Change two places we used Builtin.Executor to instead use Optional<any Actor> as an expected executor.

We want SILGen to have a simplified view of its executor and know that whenever
one sees an Actor, it is an actual actor instead of a Builtin.Executor. This
just simplifies code. Also, we should eventually have an invariant that
Builtin.Executor should only be allowed in LoweredSIL after LowerHopToExecutor
has run. But that is a change for another day.

gottesmm committed:
Test updates for previous commit.

gottesmm commented:
John asked me to change how I did this to add a new type. To make it easier, I have redone the patch without inserting the bit clearing so I can test what should be NFC from the programmers perspective (inserting the new type). I am going to test that now and then add back in the bit clearing once I get a green on that from a test perspective. So I am going to push and we are going to appear to lose the bit clearing stuff.

gottesmm commented:
... but I am going to add back in the bit clearing stuff later.

gottesmm commented:
@swift-ci smoke test

https://github.com/swiftlang/llvm-project/pull/11157

gottesmm commented:
JohnM asked me to split this up into some different PRs. I am doing this now.

gottesmm commented:
@swift-ci smoke test

gottesmm commented on include/swift/AST/TypeNodes.def:
```diff
@@ -144,7 +144,8 @@ ABSTRACT_TYPE(Builtin, Type)
   BUILTIN_GENERIC_TYPE(BuiltinFixedArray, BuiltinType)
   BUILTIN_CONCRETE_TYPE(BuiltinUnboundGeneric, BuiltinType)
-  TYPE_RANGE(Builtin, BuiltinInteger, BuiltinUnboundGeneric)
+  BUILTIN_CONCRETE_TYPE(BuiltinImplicitIsolationActor, BuiltinType)
```
Sorry, I just didn't get around to renaming it. I will fix it before this lands. I know you were thinking maybe IsolationActor... but I think that that name sounds pretty funky. We lose that it is used specifically for the implicit parameter. What about the name ImplicitActor?

gottesmm replied:
JohnM and I spoke offline... I am going to change it to ImplicitActor.

gottesmm commented on lib/SIL/IR/SILType.cpp:
```diff
@@ -103,6 +103,21 @@ SILType SILType::getEmptyTupleType(const ASTContext &C) {
+    return SILType::getEmptyTupleType(ctx);
+
+  SmallVector<TupleTypeElt, 8> tupleTypeElts;
+  auto category = elementTypes[0].getCategory();
```
I would rather pass in the category. We do have cases where we store tuples into memory so forcing it to be an object type does not compute to me.

gottesmm committed:
[sil] Add a new instruction cast_implicit_actor_to_optional_actor.

This instruction converts Builtin.ImplicitActor to Optional<any Actor>. In the
process of doing so, it masks out the bits we may have stolen from the witness
table pointer of Builtin.ImplicitActor. The bits that we mask out are the bottom
two bits of the top nibble of the TBI space on platforms that support TBI (that
is bit 60,61 on arm64). On platforms that do not support TBI, we just use the
bottom two tagged pointer bits (0,1).

By using an instruction, we avoid having to represent the bitmasking that we are
performing at the SIL level and can instead just make the emission of the
bitmasking an IRGen detail. It also allows us to move detection if we are
compiling for AArch64 to be an IRGen flag instead of a LangOpts flag.

The instruction is a guaranteed forwarding instruction since we want to treat
its result as a borrowed projection from the Builtin.ImplicitActor.

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci smoke test

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci smoke test

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci smoke test

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci smoke test

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci smoke test

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci clean smoke test

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci test windows platform

rjmccall reviewed (commented):
Okay. I didn't review the test changes in detail, but this is generally doing things the way I would expect. Broadly approved, with a few nitpicks.

rjmccall reviewed (approved):
Okay, LGTM. Please clean up the mask computation in a follow-up.

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci test windows platform

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
https://github.com/swiftlang/llvm-project/pull/11157
@swift-ci smoke test
